internal/pkg/runtime/engines/singularity/container_linux.go

```diff
@@ -1766,3 +1771,130 @@ func (c *container) prepareNetworkSetup(system *mount.System, pid int) (func() e
 
 	return nil, nil
 }
+
+func fuseDriverRunner(name string, program []string, fd int) func(system *mount.System) error {
+	return func(system *mount.System) error {
+		sylog.Debugf("Running FUSE driver for %s as %v, fd %d", name, program, fd)
+
+		fh := os.NewFile(uintptr(fd), "fd-"+name)
+		if fh == nil {
+			// this should never happen
+			return errors.New("cannot map /dev/fuse file descriptor to a file handle")
+		}
+		// the master process does not need this file descriptor
+		// after running the program, make sure it gets closed;
+		// ignore any errors that happen here
+		defer fh.Close()
+
+		// The assumption is that the plugin prepared "Program"
+		// in such a way that it's missing the last parameter
+		// and that must correspond to /dev/fd/N. Instead of
+		// making assumptions as to how many and which file
+		// descriptors are open at this point, simply assume
+		// that it's possible to map the existing file
+		// descriptor to the name number in the new process.
+		//
+		// "newFd" should be the same as "fd", but do not assume
+		// that either.
+		newFd := fh.Fd()
+		fdDevice := fmt.Sprintf("/dev/fd/%d", newFd)
+		args := append(program, fdDevice)
+		cmd := exec.Command(args[0], args[1:]...)
+
+		// Add the /dev/fuse file descriptor to the list of file
+		// descriptors to be passed to the new process.
+		//
+		// ExtraFiles is an array of *os.File, with the position
+		// of each entry determining the resulting file
+		// descriptor number. Since we are passing /dev/fd/N
+		// above, place our file handle at position N-3, so that
+		// it gets mapped to file descriptor N in the new
+		// process (the Go library will set things up so that
+		// stdin, stdout and stderr are 0, 1, and 2, so the
+		// first element of ExtraFiles gets 3).
+		cmd.ExtraFiles = make([]*os.File, newFd-3+1)
+		cmd.ExtraFiles[newFd-3] = fh
+
+		if err := cmd.Start(); err != nil {
+			sylog.Debugf("cannot start program %v: %v\n", args, err)
+			return err
+		}
+
+		return nil
+	}
+}
+
+// addFuseMount transforms the plugin configuration into a series of
+// mount requests for FUSE filesystems
+func (c *container) addFuseMount(system *mount.System) error {
+	for i, name := range c.engine.EngineConfig.GetPluginFuseMounts() {
+		var cfg struct {
+			Fuse struct {
```

Comment:
@mem Based on comment https://github.com/sylabs/singularity/pull/3679#issuecomment-501989673, wouldn't be good to have a public package with a defined FUSE config structure and also use this package could used as a doc ?

Reply:
The reason why I tried to avoid doing as suggested is because this is the configuration for a plugin. That's why it's a struct inside an anonymous struct: the outside struct is to be defined by the plugin.

I have added a global definition for the inner struct. Not happy about it, but it's there.